## Allison Wallin: Module 5

In [36]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.model_selection import GridSearchCV
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import PolynomialFeatures
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error


In [2]:
## read in data

data = pd.read_csv('/Users/allisonwallin/Desktop/Spring 2024/ATMS 523/Python/machine-learning-1-allisonwallin/homework/radar_parameters.csv')

In [3]:
data

Unnamed: 0   Zh (dBZ)  Zdr (dB)   Ldr (dB)  Kdp (deg km-1)  \
0               0  23.144878  0.418637 -41.757733        0.005395   
1               1  22.737156  0.322850 -43.772069        0.005194   
2               2  26.869826  0.330948 -43.577399        0.013385   
3               3  28.540561  0.399480 -42.139731        0.018872   
4               4  30.500127  0.543758 -39.763087        0.027438   
...           ...        ...       ...        ...             ...   
18964       18964  31.515997  0.579955 -39.244229        0.034048   
18965       18965  29.993334  0.567935 -39.399188        0.024134   
18966       18966  31.685913  0.655681 -38.375696        0.033971   
18967       18967  32.980096  0.768586 -37.166218        0.043117   
18968       18968  28.731399  0.804155 -36.795913        0.015908   

       Ah (dBZ/km)  Adr (dB/km)  R (mm/hr)  
0         0.000290     0.000012   2.393520  
1         0.000360     0.000012   3.502699  
2         0.000903     0.000030   8.627561  
3         0.001036     0.000043   8.424447  
4         0.001157     0.000064   8.189291  
...            ...          ...        ...  
18964     0.001417     0.000080  10.648020  
18965     0.001032     0.000057   7.981875  
18966     0.001165     0.000081   6.822691  
18967     0.001285     0.000105   6.801169  
18968     0.000469     0.000039   2.582440  

[18969 rows x 8 columns]

In [4]:
## drop unnamed column

data.drop(columns = 'Unnamed: 0', inplace = True)

In [5]:
data

Zh (dBZ)  Zdr (dB)   Ldr (dB)  Kdp (deg km-1)  Ah (dBZ/km)  \
0      23.144878  0.418637 -41.757733        0.005395     0.000290   
1      22.737156  0.322850 -43.772069        0.005194     0.000360   
2      26.869826  0.330948 -43.577399        0.013385     0.000903   
3      28.540561  0.399480 -42.139731        0.018872     0.001036   
4      30.500127  0.543758 -39.763087        0.027438     0.001157   
...          ...       ...        ...             ...          ...   
18964  31.515997  0.579955 -39.244229        0.034048     0.001417   
18965  29.993334  0.567935 -39.399188        0.024134     0.001032   
18966  31.685913  0.655681 -38.375696        0.033971     0.001165   
18967  32.980096  0.768586 -37.166218        0.043117     0.001285   
18968  28.731399  0.804155 -36.795913        0.015908     0.000469   

       Adr (dB/km)  R (mm/hr)  
0         0.000012   2.393520  
1         0.000012   3.502699  
2         0.000030   8.627561  
3         0.000043   8.424447  
4         0.000064   8.189291  
...            ...        ...  
18964     0.000080  10.648020  
18965     0.000057   7.981875  
18966     0.000081   6.822691  
18967     0.000105   6.801169  
18968     0.000039   2.582440  

[18969 rows x 7 columns]

### Question 1

In [6]:
## split data into 70-30 for training + testing

x = data[['Zh (dBZ)', 'Zdr (dB)', 'Ldr (dB)', 'Kdp (deg km-1)', 'Ah (dBZ/km)', 'Adr (dB/km)']]
y = data['R (mm/hr)']

x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.7)


### Question 2

In [7]:
## set model

model = LinearRegression(fit_intercept = True)


In [8]:
## train multiple linear regression dataset

model.fit(x_train, y_train)


LinearRegression()

In [10]:
## validate model with test dataset

y2_model = model.predict(x_test)

accuracy_score(y_test, y2_model)


ValueError: continuous is not supported

In [9]:
y2_model = model.predict(x_test)

## ChatGPT 3.5 suggestions for error:

mae = mean_absolute_error(y_test, y2_model)
mse = mean_squared_error(y_test, y2_model)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y2_model)

print("Mean Absolute Error:", mae)
print("Mean Squared Error:", mse)
print("Root Mean Squared Error:", rmse)
print("R-squared:", r2)


Mean Absolute Error: 0.5479454085835673
Mean Squared Error: 0.8407951570217039
Root Mean Squared Error: 0.9169488300999702
R-squared: 0.9870928134561301


In [10]:
## calculate rain rate

rate = 200 * (data['R (mm/hr)'] ** 1.6)


In [11]:
rate

0         808.145148
1        1486.190353
2        6287.107113
3        6051.963116
4        5783.943941
            ...     
18964    8803.597045
18965    5551.340603
18966    4318.760840
18967    4296.983613
18968     912.595588
Name: R (mm/hr), Length: 18969, dtype: float64

In [12]:
## repeat for calculated rain rate (baseline prediction)

x = data[['Zh (dBZ)', 'Zdr (dB)', 'Ldr (dB)', 'Kdp (deg km-1)', 'Ah (dBZ/km)', 'Adr (dB/km)']]
y = rate

# split rain rate into test and train
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.7)

# model
model = LinearRegression(fit_intercept = True)
model.fit(x_train, y_train)
y2_model = model.predict(x_test)

# calculate r^2 between the model and the calculated rate
r2 = r2_score(y_test, y2_model)
print("R-squared:", r2)

# calculate mean squared error
mse = mean_squared_error(y_test, y2_model)

# calculate root mean squared error
rmse = np.sqrt(mse)
print("Root Mean Squared Error:", rmse)


R-squared: 0.9371926390386487
Root Mean Squared Error: 5567.4924132506385


The calculated rain rate model preforms worse than the remote sensing measured rain rate. The calculated rain rate has a slightly lower R^2 at 0.94 compared to the remote sensing measurement R^2 of 0.99, so both are very high. But the RMSE is much higher for the calculated rain rate at 5567 while the remote sensing RMSE is 0.97, meaning the linear regression model outperformed the baseline prediction.

### Question 3

In [13]:
## define polynomial regression

def PolynomialRegression(degree=2, **kwargs):
    return make_pipeline(PolynomialFeatures(degree),
                         LinearRegression(**kwargs))

In [14]:
## set up grid search

param_grid = {'polynomialfeatures__degree': np.arange(10),
              'linearregression__fit_intercept': [True, False]}
              #'linearregression__normalize': [True, False]}

grid = GridSearchCV(PolynomialRegression(), param_grid, cv=7)

error:

ValueError: Invalid parameter 'normalize' for estimator LinearRegression(). Valid parameters are: ['copy_X', 'fit_intercept', 'n_jobs', 'positive'].

running for 60+ min and overheating laptop... orders reduced from 21 to 10

In [15]:
## fit grid

x = data[['Zh (dBZ)', 'Zdr (dB)', 'Ldr (dB)', 'Kdp (deg km-1)', 'Ah (dBZ/km)', 'Adr (dB/km)']]
y = data['R (mm/hr)']

x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.7)

grid.fit(x_train, y_train);


In [17]:
## pull best parameters

grid.best_params_

{'linearregression__fit_intercept': False, 'polynomialfeatures__degree': 3}

In [18]:
## pull out best model

model = grid.best_estimator_


In [20]:
## y model
y2_model = model.predict(x_test)

## calculate r^2
r2 = r2_score(y_test, y2_model)
print("R-squared:", r2)

## calculate rmse
mse = mean_squared_error(y_test, y2_model)
rmse = np.sqrt(mse)
print("Root Mean Squared Error:", rmse)


R-squared: 0.999983026028064
Root Mean Squared Error: 0.0367249390679137


The best polynomial model does outperform both the baseline and the linear regression model. It has a very high R^2 of 0.99998 while, similarly, the baseline has an R^2 of 0.94 and the linear regression model has 0.99. Its RMSE is very low at 0.037, while the linear regression model was at 0.92 and the baseline had a very high RMSE at 5567.

### Question 4

In [43]:
## assistance from ChatGPT 3.5 to work through errors + restructure code

# define data
x = data[['Zh (dBZ)', 'Zdr (dB)', 'Ldr (dB)', 'Kdp (deg km-1)', 'Ah (dBZ/km)', 'Adr (dB/km)']]
y = data['R (mm/hr)']
x_train, x_test, y_train, y_test = train_test_split(x, y, train_size = 0.7)

# hyperparameters
n_estimators_values = 200
max_depths = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, None]
min_samples_splits = [2, 5, 10]
min_samples_leafs = [1, 2, 4]
bootstrap = True

best_model = None
best_mse = float('inf')

# iterate over n_estimators
for min_samples_split in min_samples_splits:
    for min_samples_leaf in min_samples_leafs:
        for max_depth in max_depths:
            model = RandomForestRegressor(n_estimators=n_estimators_values, max_depth=max_depth, min_samples_split=min_samples_split, min_samples_leaf=min_samples_leaf, bootstrap=bootstrap)
            model.fit(x_train, y_train)
            y_pred = model.predict(x_test)
            mse = mean_squared_error(y_test, y_pred)
            if mse < best_mse:
                best_mse = mse
                best_model = model

print("Best MSE:", best_mse)
print("Best Model:", best_model)


Best MSE: 1.1849402096543675
Best Model: RandomForestRegressor(max_depth=100, min_samples_leaf=2, min_samples_split=5,
                      n_estimators=200)


In [45]:
## rmse calculation
rmse = np.sqrt(best_mse)
print("Root Mean Squared Error:", rmse)


Root Mean Squared Error: 1.0885495898921498


Based on RMSE, the random forest regressor is similar to the linear regression model (0.92) and much better than the baseline prediction (5567), but it does not do better than the polynomial model, which had an RMSE of 0.037.